packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp

```diff
@@ -1592,41 +1624,60 @@ namespace Ifpack2 {
       }
 
     public:
-      
+
+      struct ExtractAndFactorizeScalarTag {};
+      struct ExtractAndFactorizeInternalVectorTag {};
+
       KOKKOS_INLINE_FUNCTION 
       void 
-      operator() (const member_type &member) const {
+      operator() (const ExtractAndFactorizeScalarTag &, const member_type &member) const {
 	// btdm is packed and sorted from largest one 
 	const local_ordinal_type packidx = member.league_rank();
 
 	const local_ordinal_type partidx = packptr(packidx);
 	const local_ordinal_type npacks = packptr(packidx+1) - partidx;
 	const local_ordinal_type i0 = pack_td_ptr(partidx);
 	const local_ordinal_type nrows = partptr(partidx+1) - partptr(partidx);
+	Kokkos::parallel_for
+	  (Kokkos::ThreadVectorRange(member, vector_length_value), [&](const local_ordinal_type &v) {
+	    if (v < npacks) extract(member, partidx+v, v);
+	    factorize(member, i0, nrows, v, scalar_values);
+	  });
+      }
 
+      KOKKOS_INLINE_FUNCTION 
+      void 
+      operator() (const ExtractAndFactorizeInternalVectorTag &, const member_type &member) const {
+	// btdm is packed and sorted from largest one 
+	const local_ordinal_type packidx = member.league_rank();
+
+	const local_ordinal_type partidx = packptr(packidx);
+	const local_ordinal_type npacks = packptr(packidx+1) - partidx;
+	const local_ordinal_type i0 = pack_td_ptr(partidx);
+	const local_ordinal_type nrows = partptr(partidx+1) - partptr(partidx);
 	Kokkos::parallel_for
 	  (Kokkos::ThreadVectorRange(member, vector_loop_size), [&](const local_ordinal_type &v) {
             const local_ordinal_type vbeg = v*internal_vector_length;
             const local_ordinal_type vend = vbeg + internal_vector_length;
             for (local_ordinal_type vv=vbeg;vv<vend;++vv)
-              if (vv < npacks) extract(member, partidx+vv, vv);
-	    member.team_barrier();
-	    factorize(member, i0, nrows, v);
+	    if (vv < npacks) extract(member, partidx+vv, vv);
+	    factorize(member, i0, nrows, v, internal_vector_values);
 	  });
       }
 
       void run() {
 #if defined(KOKKOS_ENABLE_CUDA) && defined(IFPACK2_BLOCKTRIDICONTAINER_ENABLE_PROFILE)
         cudaProfilerStart();
```

Comment:
We should think about wrapping these CUDA profiling API calls, so that we don't have so many macros.  Also, if we have CUDA enabled, couldn't we decide at run time whether or not to enable CUDA profiling?

Reply:
That should be possible to check that in runtime using some environmental variables.

Reply:
@kyungjoo-kim We could use the existing `Tpetra::Details::Behavior` facility -- there's nothing specific about that to Tpetra.

Reply:
When I profile performance, I only need to see one particular kernel at a time. Even though I put kernel names, the profiler shows too many kernels if I use the common variable. Potentially, it also shows cuda kernels from other packages, which is too much information.

Reply:
@kyungjoo-kim `Behavior` can decide behavior based on a comma-delimited list of strings, not just a Boolean setting.  We already use this feature to restrict what classes or functions should print verbose debugging output.  We could do that for profiling as well.

Reply:
Oh... really ? Then, it will be fine. I will try to use that. Thanks.

Reply:
@kyungjoo-kim We would need to create a new environment variable and corresponding function.  Behavior tries not to do too much at run time, which is why it doesn't offer run-time look-up by environment variable.